File: domain/usecases/get_installed_packages.py
from typing import List
from domain.entities import Package
from domain.repositories import PackageRepository

class GetInstalledPackagesUseCase:
    def __init__(self, alpm_repo: PackageRepository, flatpak_repo: PackageRepository):
        self.alpm_repo = alpm_repo
        self.flatpak_repo = flatpak_repo
        self.enable_flatpak = True

    def execute(self) -> List[Package]:
        installed = []
        installed.extend(self.alpm_repo.get_installed())
        if self.enable_flatpak:
            installed.extend(self.flatpak_repo.get_installed())
        return installed

File: src-rust/src/domain/usecases/get_installed_packages.rs
use std::sync::Arc;
use crate::domain::{entities::Package, repositories::PackageRepository};
use crate::config::AppConfig;

pub struct GetInstalledPackagesUseCase {
    pub alpm_repo: Arc<dyn PackageRepository>,
    pub flatpak_repo: Arc<dyn PackageRepository>,
}

impl GetInstalledPackagesUseCase {
    pub fn new(
        alpm_repo: Arc<dyn PackageRepository>,
        flatpak_repo: Arc<dyn PackageRepository>,
    ) -> Self {
        Self { alpm_repo, flatpak_repo }
    }

    pub async fn execute(&self, config: &AppConfig) -> Vec<Package> {
        let alpm_fut = self.alpm_repo.get_installed();
        let flatpak_fut = async {
            if config.enable_flatpak {
                self.flatpak_repo.get_installed().await
            } else {
                vec![]
            }
        };

        let (alpm, flatpak) = tokio::join!(alpm_fut, flatpak_fut);
        let mut results = alpm;
        results.extend(flatpak);
        results
    }
}
